common/channelconfig/util.go

```diff
@@ -241,51 +242,72 @@ func ACLValues(acls map[string]string) *StandardConfigValue {
 
 // ValidateCapabilities validates whether the peer can meet the capabilities requirement in the given config block
 func ValidateCapabilities(block *cb.Block, bccsp bccsp.BCCSP) error {
+	cc, err := extractChannelConfig(block, bccsp)
+	if err != nil {
+		return err
+	}
+	// Check the channel top-level capabilities
+	if err := cc.Capabilities().Supported(); err != nil {
+		return err
+	}
+
+	// Check the application capabilities
+	return cc.ApplicationConfig().Capabilities().Supported()
+}
+
+// ExtractMSPIDsForApplicationOrgs extracts MSPIDs for application organizations
+func ExtractMSPIDsForApplicationOrgs(block *cb.Block, bccsp bccsp.BCCSP) ([]string, error) {
+	cc, err := extractChannelConfig(block, factory.GetDefault())
+	if err != nil {
+		return nil, err
+	}
+
+	if cc.ApplicationConfig() == nil {
+		return nil, errors.Errorf("could not get application config for channel")
+	}
+	orgs := cc.ApplicationConfig().Organizations()
+	mspids := make([]string, 0, len(orgs))
+	for _, org := range orgs {
+		mspids = append(mspids, org.MSPID())
+	}
+	return mspids, nil
+}
+
+func extractChannelConfig(block *cb.Block, bccsp bccsp.BCCSP) (*ChannelConfig, error) {
 	envelopeConfig, err := protoutil.ExtractEnvelope(block, 0)
 	if err != nil {
-		return errors.Errorf("failed to %s", err)
+		return nil, errors.Errorf("failed to %s", err)
 	}
 
 	configEnv := &cb.ConfigEnvelope{}
 	_, err = protoutil.UnmarshalEnvelopeOfType(envelopeConfig, cb.HeaderType_CONFIG, configEnv)
 	if err != nil {
-		return errors.Errorf("malformed configuration envelope: %s", err)
+		return nil, errors.Errorf("malformed configuration envelope: %s", err)
 	}
 
 	if configEnv.Config == nil {
-		return errors.New("nil config envelope Config")
+		return nil, errors.New("nil config envelope Config")
```

Comment:
Are there other types of config envelopes, if not, config is a stutter, should be `nil config envelope`, if there are can we rephrase it to `nil config envelope of type Config`

Reply:
this could be something like `envelope missing config`

Reply:
@btl5037 thanks for pointing it out. This is existing code, but should be fixed since we are changing the line anyway.